Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs

```diff
@@ -25,52 +41,94 @@ internal class TestDiscovererPytest : IPythonTestDiscoverer {
+            foreach (PytestDiscoveryResults result in discoveryResults ?? Enumerable.Empty<PytestDiscoveryResults>()) {
+                var parentMap = result.Parents.ToDictionary(p => p.Id, p => p);
+                foreach (PytestTest test in result.Tests) {
                     try {
-                        discoveryResults = JsonConvert.DeserializeObject<List<PytestDiscoveryResults>>(json);
-                    } catch (InvalidOperationException ex) {
-                        Error("Failed to parse: {0}".FormatInvariant(ex.Message));
-                        Error(json);
-                    } catch (JsonException ex) {
-                        Error("Failed to parse: {0}".FormatInvariant(ex.Message));
-                        Error(json);
+                        TestCase tc = test.ToVsTestCase(_settings.ProjectHome, parentMap);
+                        DebugInfo($"{tc.DisplayName} Source:{tc.Source} Line:{tc.LineNumber}");
+                        discoverySink?.SendTestCase(tc);
+                    } catch (Exception ex) {
+                        Error(ex.Message);
                     }
                 }
-            } catch (Exception ex) {
-                Error(ex.Message);
             }
+        }
 
-            if (discoveryResults.Any()) {
-                PyTestDiscoveryReader.ParseDiscovery(discoveryResults, discoverySink, _settings, logger);
+        private string ExecuteProcess(
+           string filename,
+           Dictionary<string, string> env,
+           IEnumerable<string> arguments,
+           string workingDirectory,
+           string pathEnv,
+           int timeoutInSeconds
+           ) {
+            using (var outputStream = new MemoryStream())
+            using (var reader = new StreamReader(outputStream, Encoding.UTF8, false, 4096, leaveOpen: true))
+            using (var writer = new StreamWriter(outputStream, Encoding.UTF8, 4096, leaveOpen: true))
+            using (var proc = ProcessOutput.Run(
+                filename,
+                arguments,
+                workingDirectory,
+                env,
+                visible: false,
+                new StreamRedirector(writer)
+            )) {
+                DebugInfo("cd " + workingDirectory);
+                DebugInfo("set " + pathEnv + "=" + env[pathEnv]);
+                DebugInfo(proc.Arguments);
+
+                if (!proc.ExitCode.HasValue) {
+                    if (!proc.Wait(TimeSpan.FromSeconds(timeoutInSeconds))) {
+                        try {
+                            proc.Kill();
+                        } catch (InvalidOperationException) {
+                            // Process has already exited
+                        }
+                        throw new TimeoutException();
+                    }
+                }
+                writer.Flush();
+                outputStream.Seek(0, SeekOrigin.Begin);
+                string data = reader.ReadToEnd();
+                return data;
```

Comment:
this should work, it's shorter and avoids writing all that stream handling code:
```
            using (var proc = ProcessOutput.Run(
                filename,
                arguments,
                workingDirectory,
                env,
                false,
                null
            )) {
                DebugInfo("cd " + workingDirectory);
                DebugInfo("set " + pathEnv + "=" + env[pathEnv]);
                DebugInfo(proc.Arguments);

                 if (!proc.ExitCode.HasValue) {
                    if (!proc.Wait(TimeSpan.FromSeconds(timeoutInSeconds))) {
                        try {
                            proc.Kill();
                        } catch (InvalidOperationException) {
                            // Process has already exited
                        }
                        throw new TimeoutException();
                    }
                }
                
                return string.Join(Environment.NewLine, proc.StandardOutputLines);
            }

```